crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs

```diff
@@ -0,0 +1,483 @@
+        "Remove extraneous blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for blank lines after end of function or class.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// class User(object):
+///     pass
+/// user = User()
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class User(object):
+///     pass
+///
+///
+/// user = User()
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E305.html)
+#[violation]
+pub struct BlankLinesAfterFunctionOrClass(pub usize);
+
+impl AlwaysAutofixableViolation for BlankLinesAfterFunctionOrClass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesAfterFunctionOrClass(blank_lines) = self;
+        format!("expected 2 blank lines after class or function definition, found ({blank_lines})")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Add missing blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for for 1 blank line between nested functions/classes definitions.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///     def inner2():
+///         pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def outer():
+///
+///     def inner():
+///         pass
+///
+///     def inner2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E306.html)
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysAutofixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// E301, E302, E303, E304, E305, E306
+pub(crate) fn blank_lines(
+    line: &LogicalLine,
+    prev_line: Option<&LogicalLine>,
+    tracked_vars: &mut BlankLinesTrackingVars,
+    indent_level: usize,
+    locator: &Locator,
+    stylist: &Stylist,
+    context: &mut LogicalLinesContext,
+) {
+    if line.is_empty() {
+        tracked_vars.blank_lines += 1;
+        tracked_vars.blank_characters += u32::try_from(line.text().len())
+            .expect("The number of blank characters should be relatively small");
```

Comment:
Nit: `len` gives you the number of bytes and not the number of blank characters. The two are different for non-ascii characters. E.g. emojis are commonly represented as 4 UTF-8 bytes, but they only form 1-2 characters.

Reply:
Fixed in 47063d4b7536e5624cf96b6cc150a62ad54a417f